Source/WebKit/UIProcess/Notifications/glib/NotificationService.cpp

```diff
@@ -0,0 +1,310 @@
+    });
+
+    return service;
+}
+
+NotificationService::NotificationService()
+{
+    const char* busName = shouldUsePortal() ? "org.freedesktop.portal.Desktop" : "org.freedesktop.Notifications";
+    const char* objectPath = shouldUsePortal() ? "/org/freedesktop/portal/desktop" : "/org/freedesktop/Notifications";
+    const char* interfaceName = shouldUsePortal() ? "org.freedesktop.portal.Notification" : "org.freedesktop.Notifications";
+    GUniqueOutPtr<GError> error;
+    m_proxy = g_dbus_proxy_new_for_bus_sync(G_BUS_TYPE_SESSION, G_DBUS_PROXY_FLAGS_DO_NOT_LOAD_PROPERTIES, nullptr, busName, objectPath, interfaceName, nullptr, &error.outPtr());
+    if (!m_proxy) {
+        g_warning("Failed to connect to notification service at %s: %s", busName, error->message);
+        return;
+    }
+
+    if (!shouldUsePortal()) {
+        GRefPtr<GVariant> capabilities = adoptGRef(g_dbus_proxy_call_sync(m_proxy.get(), "GetCapabilities", nullptr, G_DBUS_CALL_FLAGS_NONE, s_dbusCallTimeout.millisecondsAs<int>(), nullptr, &error.outPtr()));
+        if (!capabilities) {
+            g_warning("Failed to get capabilities from notification server: %s", error->message);
+            m_proxy = nullptr;
+            return;
+        }
+
+        processCapabilities(capabilities.get());
+    }
+
+    g_signal_connect(m_proxy.get(), "g-signal", G_CALLBACK(handleSignal), this);
+}
+
+void NotificationService::processCapabilities(GVariant* variant)
+{
+    GUniqueOutPtr<GVariantIter> iter;
+    g_variant_get(variant, "(as)", &iter.outPtr());
+    const char* capability;
+    while (g_variant_iter_loop(iter.get(), "&s", &capability)) {
+        if (!g_strcmp0(capability, "action-icons"))
+            m_capabilities.add(Capabilities::ActionIcons);
+        else if (!g_strcmp0(capability, "actions"))
+            m_capabilities.add(Capabilities::Actions);
+        else if (!g_strcmp0(capability, "body"))
+            m_capabilities.add(Capabilities::Body);
+        else if (!g_strcmp0(capability, "body-hyperlinks"))
+            m_capabilities.add(Capabilities::BodyHyperlinks);
+        else if (!g_strcmp0(capability, "body-images"))
+            m_capabilities.add(Capabilities::BodyImages);
+        else if (!g_strcmp0(capability, "body-markup"))
+            m_capabilities.add(Capabilities::BodyMarkup);
+        else if (!g_strcmp0(capability, "icon-multi"))
+            m_capabilities.add(Capabilities::IconMulti);
+        else if (!g_strcmp0(capability, "icon-static"))
+            m_capabilities.add(Capabilities::IconStatic);
+        else if (!g_strcmp0(capability, "persistence"))
+            m_capabilities.add(Capabilities::Persistence);
+        else if (!g_strcmp0(capability, "sound"))
+            m_capabilities.add(Capabilities::Sound);
+    }
+}
+
+bool NotificationService::showNotification(const WebNotification& notification)
+{
+    if (!m_proxy)
+        return false;
+
+    auto findNotificationByTag = [&](const String& tag) -> Notification {
```

Comment:
Better to capture only what you need, rather than everything. I think the only thing you need here is `this`, right?

Reply:
It's a lambda local to the function, but ok.